src/backup.py

```diff
@@ -0,0 +1,133 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Provides backup functionality for Synapse."""
+import logging
+from typing import Any, Optional
+
+import boto3
+import ops
+from botocore.exceptions import BotoCoreError, ClientError
+from botocore.exceptions import ConnectionError as BotoConnectionError
+from charms.data_platform_libs.v0.s3 import CredentialsChangedEvent, S3Requirer
+from ops.framework import Object
+from pydantic import BaseModel, Field, validator
+
+logger = logging.getLogger(__name__)
+
+S3_CANNOT_ACCESS_BUCKET = "Backup: S3 bucket does not exist or cannot be accessed"
+S3_INVALID_CONFIGURATION = "Backup: S3 configuration is invalid"
+BACK_UP_STATUS_MESSAGES = (S3_INVALID_CONFIGURATION, S3_CANNOT_ACCESS_BUCKET)
+
+
+class S3Parameters(BaseModel):
+    """Configuration for accessing S3 bucket.
+
+    Attributes:
+        access_key: AWS access key.
+        secret_key: AWS secret key.
+        region: The region to connect to the object storage.
+        bucket: The bucket name.
+        endpoint: The endpoint used to connect to the object storage.
+        path: The path inside the bucket to store objects.
+        s3_uri_style: The S3 protocol specific bucket path lookup type.
+    """
+
+    access_key: str = Field(alias="access-key")
+    secret_key: str = Field(alias="secret-key")
+    region: Optional[str] = Field(default=None)
+    bucket: str
+    endpoint: Optional[str]
+    path: str = Field(default="")
+    s3_uri_style: str = Field(alias="s3-uri-style", default="host")
+
+    @validator("region", always=True)
+    @classmethod
+    def check_region_or_endpoint_set(cls, region: str, values: dict[str, Any]) -> str:
+        """Validate that eather that region or endpoint are set.
+
+        Args:
+            region: region attribute
+            values: all attributes in S3 configuration
+
+        Returns:
+            value of the region attribute
+
+        Raises:
+            ValueError: if the configuration is invalid.
+        """
+        endpoint = values.get("endpoint")
+        if not region and not endpoint:
+            raise ValueError('one of "region" or "endpoint" needs to be set')
+        return region
+
+
+class SynapseBackup(Object):
+    """Class to manage Synapse backups over S3."""
```

Comment:
Does this class really manage backups? I believe its more of an observer and should perhaps be renamed to `BackupObserver` (and the module too) to distinguish between buisness and charm/event logic. There seems to be a similar split for the database. See also https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619

Reply:
Good point. I have renamed the module and class and created a new module backup for the backup related logic.

Reply:
I think at this stage (without any real business logic) I would have just kept one module and seen `can_use_bucket` as a helper method to validate the relation data, but I suppose that is debatable.